File: python/producer.py
#  ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
#  ┃ ██████ ██████ ██████       █      █      █      █      █ █▄  ▀███ █       ┃
#  ┃ ▄▄▄▄▄█ █▄▄▄▄▄ ▄▄▄▄▄█  ▀▀▀▀▀█▀▀▀▀▀ █ ▀▀▀▀▀█ ████████▌▐███ ███▄  ▀█ █ ▀▀▀▀▀ ┃
#  ┃ █▀▀▀▀▀ █▀▀▀▀▀ █▀██▀▀ ▄▄▄▄▄ █ ▄▄▄▄▄█ ▄▄▄▄▄█ ████████▌▐███ █████▄   █ ▄▄▄▄▄ ┃
#  ┃ █      ██████ █  ▀█▄       █ ██████      █      ███▌▐███ ███████▄ █       ┃
#  ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┫
#  ┃ Copyright (c) 2017, the Perspective Authors.                              ┃
#  ┃ ╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌ ┃
#  ┃ This file is part of the Perspective library, distributed under the terms ┃
#  ┃ of the [Apache License 2.0](https://www.apache.org/licenses/LICENSE-2.0). ┃
#  ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛

import random
import logging
from datetime import date, datetime
from datetime import timezone as tz
from time import sleep
import json
import taosws


logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger('main')

# =============================================================================
# TDengine connection parameters
# =============================================================================
TAOS_HOST = "localhost"                 # TDengine server host
TAOS_PORT = 6041                        # TDengine server port
TAOS_USER = "root"                      # TDengine username
TAOS_PASSWORD = "taosdata"              # TDengine password

TAOS_DATABASE = "stocks"                # TDengine database name
TAOS_TABLENAME = "stocks_values"        # TDengine table name

# =============================================================================
# Data generation parameters
# =============================================================================
INTERVAL = 250                      # seconds. insert data every INTERVAL milliseconds
NUM_ROWS_PER_INTERVAL = 250         # number of rows to insert every INTERVAL seconds
SECURITIES = [
    "AAPL.N",
    "AMZN.N",
    "QQQ.N",
    "NVDA.N",
    "TSLA.N",
    "FB.N",
    "MSFT.N",
    "TLT.N",
    "XIV.N",
    "YY.N",
    "CSCO.N",
    "GOOGL.N",
    "PCLN.N",
]
CLIENTS = ["Homer", "Marge", "Bart", "Lisa", "Maggie", "Moe", "Lenny", "Carl", "Krusty"]


class CustomJSONEncoder(json.JSONEncoder):
    """
    Custom JSON encoder that serializes datetime and date objects
    """
    def default(self, obj):
        if isinstance(obj, datetime):
            return obj.isoformat()
        elif isinstance(obj, date):
            return obj.isoformat()
        return super().default(obj)


# json.JSONEncoder.default = CustomJSONEncoder().default


def gen_data():
    """
    Generate random data
    """
    modifier = random.random() * random.randint(1, 50)
    return [{
        "timestamp": datetime.now(tz=tz.utc),
        "ticker": random.choice(SECURITIES),
        "client": random.choice(CLIENTS),
        "open": random.uniform(0, 75) + random.randint(0, 9) * modifier,
        "high": random.uniform(0, 105) + random.randint(1, 3) * modifier,
        "low": random.uniform(0, 85) + random.randint(1, 3) * modifier,
        "close": random.uniform(0, 90) + random.randint(1, 3) * modifier,
        "volume": random.randint(10_000, 100_000),
        "date": date.today(),
    } for _ in range(NUM_ROWS_PER_INTERVAL)]


def create_database(
        conn, 
        database_name: str = TAOS_DATABASE
        ) -> None:
    """
    Create a TDengine database. Drop the database if it already exists.
    """
    
    # create the database
    conn.execute(f"CREATE DATABASE IF NOT EXISTS {database_name}")
    conn.execute(f"USE {database_name}")
    logger.info(f"TDengine - Created database {database_name}")


def create_table(
        conn, 
        table_name: str = TAOS_TABLENAME
        ) -> None:
    """
    Create a TDengine table to store the data. Drop the table if it already exists.
    """
    # drop the table if it already exists
    conn.execute(f"DROP TABLE IF EXISTS {table_name}")
    # create the table
    sql = f"""
        CREATE TABLE IF NOT EXISTS `{table_name}` (
            `timestamp` TIMESTAMP,
            `ticker` NCHAR(10),
            `client` NCHAR(10),
            `open` FLOAT,
            `high` FLOAT,
            `low` FLOAT,
            `close` FLOAT,
            `volume` INT UNSIGNED,
            `date` TIMESTAMP
        )
    """
    conn.execute(sql)
    logger.info(f"TDengine - Created table {table_name}")


def insert_data(
        conn, 
        table_name: str = TAOS_TABLENAME
        ) -> None:
    """
    Insert data into the TDengine table
    """
    records = gen_data()
    
    # prepare a parameterized SQL statement
    sql = f"INSERT INTO {table_name} VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)"
    stmt = conn.statement()
    stmt.prepare(sql)
    
    # prepare the columns into their respective lists
    timestamps = [int(record['timestamp'].timestamp() * 1000) for record in records]
    tickers = [record['ticker'] for record in records]
    clients = [record['client'] for record in records]
    opens = [record['open'] for record in records]
    highs = [record['high'] for record in records]
    lows = [record['low'] for record in records]
    closes = [record['close'] for record in records]
    volumes = [record['volume'] for record in records]
    dates = [int(datetime.combine(record['date'], datetime.min.time()).timestamp() * 1000) for record in records]

    # bind the parameters and execute the statement
    stmt.bind_param([
        taosws.millis_timestamps_to_column(timestamps),
        taosws.nchar_to_column(tickers),
        taosws.nchar_to_column(clients),
        taosws.floats_to_column(opens),
        taosws.floats_to_column(highs),
        taosws.floats_to_column(lows),
        taosws.floats_to_column(closes),
        taosws.ints_to_column(volumes),
        taosws.millis_timestamps_to_column(dates),
        ]
    )
    # send the batch for insert
    stmt.add_batch()
    stmt.execute()
    logger.debug(f"TDengine - Wrote {len(records)} rows to table {table_name}")



def main():
    """
    Create a TDengine client, create the database and table, and insert data into the table
    """
    # create a tdengine websocket client
    conn = taosws.connect(host=TAOS_HOST, port=TAOS_PORT, user=TAOS_USER, password=TAOS_PASSWORD)

    # create the database and table
    create_database(conn, database_name=TAOS_DATABASE)
    create_table(conn, table_name=TAOS_TABLENAME)
    
    progress_counter = 0
    logger.info(f"Inserting data to TDengine @ interval={INTERVAL:d}ms...")
    try:
        while True:
            insert_data(conn, table_name=TAOS_TABLENAME)
            progress_counter += 1
            print('.', end='' if progress_counter % 80 else '\n', flush=True)
            sleep((INTERVAL / 1000.0))
    except KeyboardInterrupt:
        logger.info(f"Shutting down...")


if __name__ == "__main__":
    main()

File: node/src/producer.js
import taos from "@tdengine/websocket";


/* -------------------------------------------------------------
    * TDengine connection information
    * Change the connection type to 'cloud' or 'docker' as needed.
 ----------------------------------------------------------------
 */
const TAOS_CONNECTION_TYPE = 'cloud'; // 'cloud' or 'local' (docker)


let TAOS_CONNECTION_URL, TAOS_USER, TAOS_PASSWORD;

const TAOS_DATABASE = 'webinar';
const TAOS_TABLENAME = 'market';

if (TAOS_CONNECTION_TYPE === 'cloud') {
    // TDengine Cloud connection
    TAOS_CONNECTION_URL = process.env.TDENGINE_CLOUD_URL || `wss://YOUR_CLOUD_REGION.cloud.tdengine.com?token=YOUR_TDENGINE_CLOUD_TOKEN`;
} else if (TAOS_CONNECTION_TYPE === 'local') {
    // TDengine local docker connection
    TAOS_CONNECTION_URL = 'ws://localhost:6041';
    TAOS_USER = 'root';
    TAOS_PASSWORD = 'taosdata';
}

/* -------------------------------------------------------------
    * Data generation information
 ----------------------------------------------------------------
 */

// Data generation constants
const MAX_RUN_DURATION = 60 * 60 * 1000;    // 1 hour in milliseconds
const INTERVAL = 250;                       // interval in milliseconds
const NUM_ROWS_PER_INTERVAL = 100;          // rows generated per interval


const COMPANY_METADATA = [
    { ticker: "AAPL.N", sector: "Information Technology", state: "CA", index_fund: ["S&P 500","NASDAQ 100","DJIA","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [169.21, 260.10], avg_volume: 59450000 },
    { ticker: "AMZN.N", sector: "Consumer Discretionary", state: "WA", index_fund: ["S&P 500","NASDAQ 100","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [151.61, 242.52], avg_volume: 50520000 },
    { ticker: "NVDA.N", sector: "Information Technology", state: "CA", index_fund: ["S&P 500","NASDAQ 100","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [86.62, 153.13], avg_volume: 296000000 },
    { ticker: "TSLA.N", sector: "Consumer Discretionary", state: "TX", index_fund: ["S&P 500","NASDAQ 100","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [167.41, 488.54], avg_volume: 123000000 },
    { ticker: "MSFT.N", sector: "Information Technology", state: "WA", index_fund: ["S&P 500","NASDAQ 100","DJIA","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [344.79, 468.35], avg_volume: 23000000 },
    { ticker: "GOOGL.N", sector: "Communication Services", state: "CA", index_fund: ["S&P 500","NASDAQ 100","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [142.66, 208.70], avg_volume: 38320000 },
    { ticker: "JPM.N", sector: "Financials", state: "NY", index_fund: ["S&P 500","DJIA","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [190.88, 280.25], avg_volume: 11380000 },
    { ticker: "V.N", sector: "Financials", state: "CA", index_fund: ["S&P 500","DJIA","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [252.70, 366.54], avg_volume: 7600000 },
    { ticker: "DIS.N", sector: "Communication Services", state: "CA", index_fund: ["S&P 500","DJIA","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [80.10, 118.63], avg_volume: 11000000 },
    { ticker: "WMT.N", sector: "Consumer Staples", state: "AR", index_fund: ["S&P 500","DJIA","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [59.44, 105.30], avg_volume: 25300000 },
    { ticker: "PFE.N", sector: "Health Care", state: "NY", index_fund: ["S&P 500","DJIA","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [20.92, 31.54], avg_volume: 55100000 },
    { ticker: "ORCL.N", sector: "Information Technology", state: "TX", index_fund: ["S&P 500","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [114.55, 198.31], avg_volume: 11200000 },
    { ticker: "NFLX.N", sector: "Communication Services", state: "CA", index_fund: ["S&P 500","NASDAQ 100","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [587.04, 1164.00], avg_volume: 4100000 },
    { ticker: "INTC.N", sector: "Information Technology", state: "CA", index_fund: ["S&P 500","NASDAQ 100","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [17.67, 37.16], avg_volume: 104400000 },
    { ticker: "ADBE.N", sector: "Information Technology", state: "CA", index_fund: ["S&P 500","NASDAQ 100","Russell 1000","S&P 400 MidCap","Wilshire 5000"], price_range: [332.01, 587.75], avg_volume: 4100000 },
];

const CLIENTS = [
    "BlackRock", "Vanguard", "State Street",
    "Fidelity", "Goldman Sachs", "Morgan Stanley",
    "Citadel Securities", "Bridgewater", "Berkshire Hathaway",
];


/**
 * Returns the current date and time in ISO 8601 format with a random jitter of ±5 seconds.
 */
function _nowISO() {
    const offset = (Math.random() * 10) - 5; // jitter +/-5 sec
    return new Date(Date.now() + offset * 1000).toISOString();
}


/**
 * Generates a random number following a normal (Gaussian) distribution using the Box-Muller transform.
 */
function _randomDistribution(mean, stddev) {
    let u = 0, v = 0;
    while(u === 0) u = Math.random();
    while(v === 0) v = Math.random();
    let num = Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
    return num * stddev + mean;
}


/**
 * Generates an array of random market data objects for a given number of rows.
 *
 * @param {number} [num_rows=NUM_ROWS_PER_INTERVAL] - The number of data rows to generate.
 * @returns {Object[]} Array of generated market data objects.
 */
function generateData(num_rows = NUM_ROWS_PER_INTERVAL) {
    const results = [];
    for (let i = 0; i < num_rows; i++) {
        // --- random company
        const metaIdx = Math.floor(Math.random() * COMPANY_METADATA.length);
        const meta = COMPANY_METADATA[metaIdx];

        // --- price bars
        const [lowRange, highRange] = meta.price_range;
        const open = +(Math.random() * (highRange - lowRange) + lowRange).toFixed(2);
        const high = +(open * (1 + Math.random() * 0.03)).toFixed(2);
        const low = +(open * (0.97 + Math.random() * 0.03)).toFixed(2);
        const close = +(Math.random() * (high - low) + low).toFixed(2);

        // --- market metrics
        const avgVol = meta.avg_volume;
        const volume = Math.max(1, Math.round(_randomDistribution(avgVol, avgVol * 0.15)));
        const lotSize = Math.floor(Math.random() * (500 - 50)) + 50;
        const trade_count = Math.floor(volume / lotSize);
        const notional = +(close * volume).toFixed(2);

        // --- dimensions & timestamps
        const index_fund = meta.index_fund[Math.floor(Math.random() * meta.index_fund.length)];
        const client = CLIENTS[Math.floor(Math.random() * CLIENTS.length)];
        const country = "United States";
        const trade_date = (new Date()).toISOString().slice(0, 10);
        const last_update = _nowISO();

        results.push({
            ticker: meta.ticker,
            sector: meta.sector,
            state: meta.state,
            index_fund,
            open,
            high,
            low,
            close,
            volume,
            trade_count,
            notional,
            client,
            country,
            trade_date,
            last_update
        });
    }
    return results;
}


/* -------------------------------------------------------------
    * TDengine connection and data insertion functions
 ----------------------------------------------------------------
 */


/**
 * Connect to a TDengine database using WebSocket APIs.
 */
async function taosCreateConnection(
    url = TAOS_CONNECTION_URL, 
    user = (TAOS_USER || null), 
    password = (TAOS_PASSWORD || null)
) {
    try {
        let conf, conn;
        if (TAOS_CONNECTION_TYPE === 'cloud') {
            conf = new taos.WSConfig(url);
            conf.setTimeOut(15_000);
            conn = await taos.sqlConnect(conf);
            console.log(`Connected to TDengine Cloud at ${url} successfully.`);
            return conn;
        } else if (TAOS_CONNECTION_TYPE === 'local') {
            conf = new taos.WSConfig(url);
            conf.setUser(user);
            conf.setPwd(password);
            conn = await taos.sqlConnect(conf);
            console.log(`Connected to local TDengine at ${url} successfully.`);
            return conn;
        } else {
            throw new Error("Invalid TAOS_CONNECTION_TYPE. Please set it to 'cloud' or 'docker'.");
        }
    } catch (err) {
        console.log(`Failed to connect to ${url}, ErrCode: ${err.code}, ErrMessage: ${err.message}`);
        if (TAOS_CONNECTION_TYPE === 'local') {
            console.log("Please run `/docker.sh` to start the TDengine docker container.");
        }
        process.exit(1);
    }
}


/**
 * Create a TDengine database and table.
 */
async function taosCreateDatabase(conn, databaseName = TAOS_DATABASE) {
    try {
        // create the database with a 1-year retention policy
        await conn.exec(`CREATE DATABASE IF NOT EXISTS ${databaseName} KEEP 365 DURATION 10 BUFFER 16 WAL_LEVEL 1;`);
        // switch to the database
        await conn.exec(`USE ${databaseName};`);
        console.log(`Database ${databaseName} created successfully.`);
    } catch (err) {
        console.error(`Failed to create database ${databaseName}, ErrCode: ${err.code}, ErrMessage: ${err.message}`);
        throw err;
    }
}


/**
 * Create a TDengine table for market data matching the generateData() schema.
 * Uses VARCHAR for all string columns.
 */
async function taosCreateTable(conn, databaseName = TAOS_DATABASE, tableName = TAOS_TABLENAME) {
    try {
        // Drop the table if it already exists
        await conn.exec(`USE ${databaseName};`);
        await conn.exec(`DROP TABLE IF EXISTS ${databaseName}.${tableName};`);

        // Create the table with columns matching generateData()
        const sql = [
            "CREATE TABLE IF NOT EXISTS `" + databaseName + "`.`" + tableName + "` (",
            "    `ts` TIMESTAMP,",
            "    `ticker` VARCHAR(16),",
            "    `sector` VARCHAR(32),",
            "    `state` VARCHAR(8),",
            "    `index_fund` VARCHAR(32),",
            "    `open` FLOAT,",
            "    `high` FLOAT,",
            "    `low` FLOAT,",
            "    `close` FLOAT,",
            "    `volume` BIGINT,",
            "    `trade_count` INT,",
            "    `notional` DOUBLE,",
            "    `client` VARCHAR(32),",
            "    `country` VARCHAR(32),",
            "    `trade_date` TIMESTAMP,",
            "    `last_update` TIMESTAMP",
            ");"
        ].join('\n');
        await conn.exec(sql);
        console.log(`TDengine - Created table ${databaseName}.${tableName} for market data`);
    } catch (err) {
        console.error(`Failed to create table ${tableName}, ErrCode: ${err.code}, ErrMessage: ${err.message}`);
        throw err;
    }
}


/**
 * Insert data into a TDengine table using the WebSocket API and batch API for performance.
 * Inserts market data rows as generated by generateData().
 */
async function toasInsertData(conn, data, databaseName = TAOS_DATABASE, tableName = TAOS_TABLENAME) {
    try {
        const sql = `
            INSERT INTO ${databaseName}.${tableName} 
            (ts, ticker, sector, state, index_fund, open, high, low, close, volume, trade_count, notional, client, country, trade_date, last_update)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        `;

        let stmt = await conn.stmtInit();
        await stmt.prepare(sql);

        let bindParams = stmt.newStmtParam();
        bindParams.setTimestamp(data.map(row => new Date(row.last_update).getTime()));
        bindParams.setVarchar(data.map(row => row.ticker));
        bindParams.setVarchar(data.map(row => row.sector));
        bindParams.setVarchar(data.map(row => row.state));
        bindParams.setVarchar(data.map(row => row.index_fund));
        bindParams.setFloat(data.map(row => row.open));
        bindParams.setFloat(data.map(row => row.high));
        bindParams.setFloat(data.map(row => row.low));
        bindParams.setFloat(data.map(row => row.close));
        bindParams.setInt(data.map(row => row.volume));
        bindParams.setInt(data.map(row => row.trade_count));
        bindParams.setDouble(data.map(row => row.notional));
        bindParams.setVarchar(data.map(row => row.client));
        bindParams.setVarchar(data.map(row => row.country));
        bindParams.setTimestamp(data.map(row => row.trade_date));
        bindParams.setTimestamp(data.map(row => new Date(row.last_update).getTime()));

        await stmt.bind(bindParams);
        await stmt.batch();
        await stmt.exec();
        console.log(`Inserted ${data.length} rows into table: ${databaseName}.${tableName}`);
    } catch (err) {
        console.error(`Failed to insert rows into ${tableName}, ErrCode: ${err.code}, ErrMessage: ${err.message}`);
        throw err;
    }
}


/**
 * Main function to generate and insert random power line data into a TDengine database at regular intervals.
 * 
 * We establish a connection to the TDengine database, creates the necessary database and table,
 * and then starts a loop to generate and insert data at a specified interval. The loop runs until a maximum
 * duration is reached, at which point the connection is closed, and the process exits.
 * 
 * Workflow:
 * 1. Create a TDengine connection >> `taosCreateConnection()`.
 * 2. Create the demo database and table >> `taosCreateDatabase()`, `taosCreateTable()`.
 * 3. Start a data generation and insertion timer loop >> `generateData()`, `toasInsertData()`.
 * 
 */
async function main() {

    try {
        // create a connection to the TDengine database
        let conn = await taosCreateConnection();
        // create the database and table
        TAOS_CONNECTION_TYPE == 'local' && await taosCreateDatabase(conn);
        await taosCreateTable(conn);
        await conn.exec(`USE ${TAOS_DATABASE};`);

        // start the data generation and insertion loop
        const startTime = Date.now();
        console.log("Starting data generation and insertion loop...");

        // insert data at a regular interval -- 250ms default
        const intervalId = setInterval(async () => {
            try {
                // exit if the max run duration is reached -- 1 hour default
                if ((Date.now() - startTime) >= MAX_RUN_DURATION) {
                    console.log("Max run duration reached. Exiting...");
                    clearInterval(intervalId);
                    await conn.close();
                    await taos.destroy();
                    process.exit(0);
                }
                // generate and insert data
                let data = generateData();
                await toasInsertData(conn, data);
            } catch (err) {
                console.error("Error during data insertion:", err);
            }
        }, INTERVAL);
    } catch (err) {
        console.log("Demo Failed! ErrCode: " + err.code + ", ErrMessage: " + err.message);
    }
}

// Run the main function
try {
    await main();
    await taos.destroy();
    
    // Handle graceful shutdown on SIGINT (Ctrl+C)
    process.on('SIGINT', async () => {
        console.log("Received SIGINT. Closing connection...");
        await taos.destroy();
        process.exit(0);
    });
} catch (err) {
    console.error("Fatal error:", err);
}
